docs/proposed/pub-sub.rst

```diff
@@ -0,0 +1,203 @@
+
+Write Key -- first 16 bytes of SH256d("allmydata_mutable_privkey_to_writekey_v1" + Signature Key)
+
+Signature Key -- private part of SDMF keypair
+
+Verification Key -- public part of SDMF keypair
+
+SDMF keypair -- random public/private keypair (RSA 2048-bit)
+
+
+Proposed Solution
+=================
+
+Anything wanting a "broadcast" or "pub-sub" like pattern can implement
+it using the following mechanism:
+
+ - the publisher creates a mutable directory, yielding a write-cap, "W"
+ - anyone wishing to subscribe is given a read-cap, "R"
+   - anyone possessing a valid read-cap "R" (including the publisher)
+     can obviously "invite" further clients
+   - nobody else may subscribe
+ - to add information, the publisher adds a new file to the mutable
+   directory
+ - old information can be removed (or not) as the publisher wishes
+
+Without further changes to Tahoe-LAFS, this can already be achieved:
+clients can "poll" the read-capability and determine if anything new
+exists.
+
+
+Improvements Over Polling
+-------------------------
+
+The above mechanism would scale somewhat poorly as the number of
+"broadcast" capabilities increased. It would be better if the
+storage-server could *tell* clients when a mutable Slot has changed.
+
+Storage servers implement a new feature, based on WebSockets. See also
+:ref:`http-storage-node-protocol` with which this protocol aims to be
+compatible. A single WebSocket endpoint exists:
+
+ - <server>/v1/mutable_updates
+
+After connecting to this endpoint, a client may send any number of
+messages (encoded using JSON) asking for updates to mutable
+Slots:
+
+    {
+        "tahoe-mutable-notification-version": 1,
+        "subscribe": [
+            "storage-index-0",
+            "storage-index-1",
+        }
+    }
+
+..where "storage-index-0" corresponds to an actual Storage Index of an
+existing mutable file. A client computes this from "R" (a read
+capability). The client must keep a mapping of `Storage Index ->
+read-capability` so it can match subsequent notifications.
+
+The server replies with a message like:
+
+    {
+        "tahoe-mutable-notification-version": 1,
+        "status": {
+            "storage-index-0": true,
+            "storage-index-1": "some kind of error message"
+        }
+    }
+
+That is, a "status" dict containing each requested Slot's
+storage-index with either "true" if the subscribe was successful or a
+string with an error-message if it was not possible to subscribe
+(e.g. the capability wasn't a mutable one, or wasn't found on this
+server). Every storage-index requested in the initial list will have
+an entry in this dict. The client should remember a map of all
+*successful* subscriptions back to the corresponding read-capability.
+
+TODO:
+
+ - for a given Slot, does a client just ask every storage-server for
+   the corresponding Storage Index? (are there information-disclosure
+   issues here?)
```

Comment:
> Relatedly, I wonder how a client decides which storage servers on which to subscribe to updates. 

Since this is for updates to existing mutables, we'll know for sure, no? (That is, when we ask for the updates, that sever will know if it has those or not). As an optimization we could say something like "clients MAY choose to only connect to N or more servers for the mutables they're interested in".

And yes, we should include more words about N / K issues I think (that is, the client must contact at least N (and at most K or "happy") servers and only accept that an update "has happened" when N say they've gotten it).

Still, for a version 0 I think the total bandwidth will still be a lot lower if we just connect to every storage-server we know about (versus contacting N storage-severs that have a share of that mutable every X minutes). I guess the proposal could be strengthened by including some math about this point ;)

Reply:
I've now included words about coalescing (and a `maximum-delay` option). I haven't addressed any N / K stuff (yet).

Reply:
I found this ticket: https://tahoe-lafs.org/trac/tahoe-lafs/ticket/2555

Reply:
Let's steal it.